lib/kafka/consumer.rb

```diff
@@ -94,6 +94,7 @@ def subscribe(topic, default_offset: nil, start_from_beginning: true, max_bytes_
     # @return [nil]
     def stop
       @running = false
+      @cluster.disconnect
```

Comment:
You know that this happens in a different thread when catching a signal right?

Reply:
Are you running the consumer outside the main thread? I believe signal handlers run in the main thread, but I may be wrong.

Reply:
@dasch yes ruby executes the signal handler in the same thread as the parent but I don't execute consumer that way as I have multi threaded environment and this change makes it unsafe to stop the execution (it was ok without that because of a fetch loop)

Reply:
The problem is that during the fetch loop, the thread will block even after `stop` is called – that's why a more aggressive shutdown is required.

The consumer isn't designed to be thread safe – in this case, can you push a message into the thread running the consumer and have that thread call `stop` itself rather than calling it from another thread? That would of course mean you couldn't use `each_message` or `each_batch`.

Reply:
> That would of course mean you couldn't use each_message or each_batch.
So I would be left with the non-merged manual api for consumption, right?

Reply:
I might have to push this until after the new consumer design has been implemented – when the fetches are being done in separate threads it will be easier to manage this.

Reply:
You mean fetches for separate topics out out the same consumer group, right?

Reply:
Each consumer would have a separate fetcher thread _per broker connection_, so we can parallelize the fetches. The fetches would still have to be interruptible though, unless I just decide to kill the threads on shutdown...

Reply:
@dasch yeah I will just drop celluloid in favor of threads and signaling with ```handle_interrupt``` and will just execute shutdown out of each thread natively. But the #424 still remains in that case.

Reply:
I still think that #369 is the right level of abstraction for that use case. Can you try to test out the API?

Reply:
@dasch yes - we agreed that I will test it BUT it still adds a lot of boilerplate to cover just a single side case. Nonetheless you can count on me in that matter (just needs a bit more time to adapt to it)

Reply:
👍 

I think the `each_*` should be the "high-level" consumer APIs in ruby-kafka, with `poll` being used by other libraries/frameworks and people with special needs (a small minority).

Reply:
@mensfeld thinking about this, I think it shouldn't be a problem that `close` is called from a different thread – the write to `@cancel_writer` is thread safe, as it's just a pipe.

Reply:
Can you verify that it's a problem in an application of yours?

Reply:
@dasch it's no longer a problem as I redesigned my consumers to be 100% thread safe ;) the trap context changes the internal state and this propages the stop:

```ruby
def consumer_loop
  super do
    if Karafka::App.stopped?
      stop
    else
      yield
    end
  end
end
```

Reply:
:+1: